### PR TITLE
Variable GPU Histogram of oriented gradients

### PR DESCRIPTION
**Pull request by ManuelFreudenreich**
Changed hog to work with variable parameters, changed the hog sample to test it with more options, added comments and tests.

### CODE REVIEW

**Comment by vpisarev** [comment]
@jet47, I realize this thing may be unknown to you, but since I do not know CUDA at all, could you at least check it for overall correctness

File: modules/cudaobjdetect/test/test_objdetect.cpp [hunk 1]
```diff
@@ -269,8 +269,241 @@ INSTANTIATE_TEST_CASE_P(detect, CalTech, testing::Combine(ALL_DEVICES,
...
+    gpu_hog->setHitThreshold(0);
+    gpu_hog->setWinStride(win_stride);
+    gpu_hog->setScaleFactor(1.05);
+    gpu_hog->setGroupThreshold(8);
+    gpu_hog->compute(_img, d_img);
+
+    vector<float> gpu_desc_vec;
+    ASSERT_TRUE(gpu_desc_vec.empty());
+    cv::Mat R(d_img);
+    std::cout << "R (csv)     = " << std::endl << R << std::endl << std::endl;
```
Comment (jet47) [inline modules/cudaobjdetect/test/test_objdetect.cpp]: Please remove all this outputs from the test, or at least make it optional and disabled by default.

**Comment by ManuelFreudenreich** [comment]
I removed the outputs from the test.

**Comment by jet47** [comment]
:+1:

**Comment by GabrieleDalmazzone** [comment]
Hi everyone,
I have verified that this new implementation with the following set:
*par = 16;
imgSize = (256, 256);
winSize = imgSize;
blockSize = (2 * par, 2 * par);
blockStride = (par, par);
cellSize = (par, par); 
nbins = 9; 
 winSigma = 16; 
 threshold_L2hys = 0.2;
gamma_correction=false;*
returns values that, even if they have sense, are not compatible with the one provided by the standard cpu implementation with same parameters (differences up to 50%)

This doesn’t happen if I modify the previous settings with the only change
*par = 8*
which was the only supported parameters setting before this commit.

I'm trying to track down the problem in the code, but without success...

**Pull request merged**